### ISSUE TITLE
custom function too slow, how to be as  fast as native?

### ISSUE THREAD

**Issue by laohur** [pytorch/pytorch#42777]
### conext
ubuntu18.04.1 64 cuda11 pytorch1.6 

```
class Relu0(Function):
    @staticmethod
    def forward(ctx, input,d=1):
        ctx.save_for_backward(input)  # save input for backward pass
        x=torch.clone(input)
        x[x<0]=0
        return x

    @staticmethod
    def backward(ctx, grad_output,d=1):
        input, = ctx.saved_tensors # restore input from context
        grad_output[input<0]= 0
        return grad_output

# more faster , not enough
def relu(x):
    return (x>0)*x

# clamp fastest , any more?

```

### benchmark
Relu0.apply cost 10 times of torch.nn.functional.relu

above way seems not use accelerate library.
which operations can be as fast as native?

cc @VitalyFedyunin @ngimel

**Comment by robieta** [comment]
In general this question is better suited to the [PyTorch forums](https://discuss.pytorch.org/), however I put together a brief benchmark to demonstrate some of the high level principles at work:

<details>

<summary> Code </summary>

```
import torch
from torch.utils._benchmark import Timer, Compare

torch.random.manual_seed(0)

@torch.jit.script
def index_relu0(x):
    x = x.clone()
    x[x<0]=0
    return x

@torch.jit.script
def mask_relu0(x):
    return (x>0)*x


_RELU_IMPL = {
    "index_relu0": index_relu0,
    "mask_relu0": mask_relu0,
    "nn_relu": torch.nn.ReLU(),
}

_SIZES = {
    "1024": (1024, 1),
    "(1024, 1024)": (1024, 1024),
    "(8192, 8192)": (8192, 8192),
}


results = []
for num_threads in [1, 4, 16]:
    for size_str, size in _SIZES.items():
        x_standard = torch.rand(size)
        x_strided = torch.rand(tuple(i * 4 for i in size))[::4, ::4]
        x_layout_transposed = torch.rand(size).t().contiguous().t()
        x_cfgs = [
            ("", x_standard),
            (" (strided x4)", x_strided),
            (" (transposed)", x_layout_transposed)
        ]
        if size[1] == 1:
            x_cfgs.pop()
        for x_suffix, x in x_cfgs:
            for fn_name, fn in _RELU_IMPL.items():
                stmt = f"{fn_name}(x)"
                timer = Timer(
                    stmt,
                    globals={"x": x, fn_name: fn},
                    description=size_str,
                    label=f"ReLU{x_suffix}",
                    sub_label=stmt,
                    num_threads=num_threads,
                )
                results.append(timer.blocked_autorange(min_run_time=1))

cmp = Compare(results)
cmp.trim_significant_figures()
cmp.print()
```

</details>

<details>

<summary> Results </summary>

```
[---------------------------- ReLU ---------------------------]
                      |  1024  |  (1024, 1024)  |  (8192, 8192)
1 threads: ----------------------------------------------------
      index_relu0(x)  |   43   |      1300      |     320000
      mask_relu0(x)   |   24   |      4400      |     500000
      nn_relu(x)      |    8   |       380      |     210000
4 threads: ----------------------------------------------------
      index_relu0(x)  |   44   |       900      |     100000
      mask_relu0(x)   |   24   |       384      |     100000
      nn_relu(x)      |    8   |       100      |      60000
16 threads: ---------------------------------------------------
      index_relu0(x)  |   59   |       730      |      80000
      mask_relu0(x)   |   24   |       190      |      72000
      nn_relu(x)      |    8   |        50      |      30000

Times are in microseconds (us).


[--------------------- ReLU (strided x4) ---------------------]
                      |  1024  |  (1024, 1024)  |  (8192, 8192)
1 threads: ----------------------------------------------------
      index_relu0(x)  |   48   |      2200      |     420000
      mask_relu0(x)   |   30   |      9600      |     700000
      nn_relu(x)      |   11   |      1600      |     310000
4 threads: ----------------------------------------------------
      index_relu0(x)  |   47   |       997      |     200000
      mask_relu0(x)   |   29   |      1100      |     200000
      nn_relu(x)      |   11   |       352      |     110000
16 threads: ---------------------------------------------------
      index_relu0(x)  |   49   |       790      |      93000
      mask_relu0(x)   |   30   |       300      |      99000
      nn_relu(x)      |   11   |       140      |      42000

Times are in microseconds (us).


[---------------- ReLU (transposed) -----------------]
                      |  (1024, 1024)  |  (8192, 8192)
1 threads: -------------------------------------------
      index_relu0(x)  |      3800      |    1457000
      mask_relu0(x)   |     11000      |    2534000
      nn_relu(x)      |       371      |     190000
4 threads: -------------------------------------------
      index_relu0(x)  |      2300      |     440000
      mask_relu0(x)   |      2000      |     690000
      nn_relu(x)      |       130      |      70000
16 threads: ------------------------------------------
      index_relu0(x)  |      1100      |     220000
      mask_relu0(x)   |       500      |     300000
      nn_relu(x)      |        44      |      29000

Times are in microseconds (us).
```

</details>

The thing to remember about ReLU is that it has a low [arithmetic intensity](https://en.wikipedia.org/wiki/Roofline_model#Arithmetic_intensity), so managing memory is quite important to get good performance. Your first function (which I call `index_relu0`) uses indexing which requires a separate value to be consulted (and pointer dereference) compared to just directly using the value when zeroing an element. The second suffers from the fact that compare+cast+multiply is more expensive than max. If we look at the following bit of C++ in [Godbolt](https://godbolt.org/) (Clang and `-O2`):

<details>

<summary> C++: </summary>

```
#include <algorithm>

double non_neg(double x){
    return std::max(x, 0.0);
}

double non_neg_multiply(double x){
    return x * (double)(x >= 0);
}
```

</details>

<details>

<summary> We get the following assembly: </summary>

```
non_neg(double):                            # @non_neg(double)
        xorpd   xmm1, xmm1
        maxsd   xmm1, xmm0
        movapd  xmm0, xmm1
        ret
.LCPI1_0:
        .quad   4607182418800017408     # double 1
non_neg_multiply(double):                  # @non_neg_multiply(double)
        xorpd   xmm1, xmm1
        cmplesd xmm1, xmm0
        movsd   xmm2, qword ptr [rip + .LCPI1_0] # xmm2 = mem[0],zero
        andpd   xmm2, xmm1
        mulsd   xmm0, xmm2
        ret
```

</details>

The specifics aren't critical; the main takeaway is that in the first case there is a direct max instruction `maxsd`, and in the second we have 50% more instructions because the function is formulated as a math rather than bit-manipulation task. 

Also, in both cases, the homebrew ReLU functions create an intermediate mask Tensor, and this consumes extra memory bandwidth which is a problem due to the low arithmetic intensity.

Lastly, the `nn.ReLU` implementation has two additional bit of engineering. First, it uses vectorized instructions to get some additional performance out of the CPU (that part of the source code has some build magic to get SSE/AVX instructions):
  
https://github.com/pytorch/pytorch/blob/91f3114fc1350dd2e1797e6edc5c61c44f2ffecd/aten/src/ATen/native/cpu/Activation.cpp#L90-L94

and second, it uses an internal helper called [TensorIterator](https://labs.quansight.org/blog/2020/04/pytorch-tensoriterator-internals/):

https://github.com/pytorch/pytorch/blob/91f3114fc1350dd2e1797e6edc5c61c44f2ffecd/aten/src/ATen/native/Activation.cpp#L362

The first helps with the overall baseline (`nn.relu` is basically always at least 2x as fast as the homebrew implementations) in addition to the use of `max` rather than `multiply` and the lack of intermediate allocations discussed earlier, while TensorIterator helps with the non-contiguous cases by doing clever layout analysis and is the reason why `nn.relu` is almost unaffected by the transpose layout while the others see severely degraded performance due to factors such as high cache miss rates.

Hope this was helpful.
